lib/THC/THCReduceApplyUtils.cuh

```diff
@@ -125,7 +237,7 @@ __device__ T reduceBlockWithNThreadLocalReductions(T *smem,
     local = reduceOp(local, next);
   }
 
-  return reduceBlock<T, ReduceOp>(smem, blockDim.x < numVals ? blockDim.x : numVals, local, reduceOp, init);
+  return reduceBlock<T, ReduceOp>(smem, THCCeilDiv(numVals, N), local, reduceOp, init);
```

Comment:
For the new code: `numVals` is the overall slice size for the reduction. So If `numVals` is 512 and `N` = 2 then the first 256 threads in the block have valid input values. For `reduceBlock`, numVals represents the number of threads whose values should be considered valid. So in the above example, the first 256 threads have values that should be reduced, hence the division.

The old code was doing something incorrect. However, because the local reduction uses init, it still succeeded.